### PR TITLE
planner: standalone implement for `PlanContext`

### LINKED ISSUE #51236: support to implement `PlanContext` in separate package outside session

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #51236

### What changed and how does it work?

In this PR, we introduced a standalone implement for `PlanContext` without writing all implementations in `session` package. Now, the `sessionctx.Context` is not required to implement all methods in `PlanContext` and that makes the `PlanContext` more flexible and easy to add and delete methods without affecting `sessionctx.Context`.

`sessionctx.Context` also provides a new method `GetPlanCtx` to return a plan context related with session. Further more it is more clear to identify in which case plan context are used when executing a SQL.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

File: pkg/planner/context/context.go [hunk 1]
```diff
@@ -49,4 +49,14 @@ type PlanContext interface {
...
 	// HasDirtyContent checks whether there's dirty update on the given table.
 	HasDirtyContent(tid int64) bool
+	// AdviseTxnWarmup advises the txn to warm up.
+	AdviseTxnWarmup() error
 }
+
+// EmptyPlanContextExtended is used to provide some empty implementations for PlanContext.
+// It is used by some mock contexts that are only required to implement PlanContext
+// but do not care about the actual implementation.
+type EmptyPlanContextExtended struct{}
```
Comment (YangKeao) [inline pkg/planner/context/context.go:59]: It's used to provide implementation for "optional" methods in `PlanContext`, right? Cool 👍 .

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/planner/context/context.go (hunk 1 above)
Comment (lcwangchao) [reply pkg/planner/context/context.go:59]: Yes, currently it is used by mock contextes

**Review by lance6716 — APPROVED** [review]

**Pull request merged**